# 아이템 39. 명명 패턴보다 애너테이션을 사용하라

## 명명패턴
전통적으로 도구나 프레임워크가 특별히 다뤄야 할 프로그램 요소에는 구분되는 명명 패턴을 적용했다.
ex) 과거 junit은 테스트 메서드 이름을 test로 시작하게끔 했다.
### 단점
#### 오타가 나면 안된다.
- junit에서 실수로 tsetSafety 같이 오타를 내면 이 테스트를 무시함
#### 올바른 프로그램 요소에서만 사용되리라 보증할 방법 없다.
- 메서드가 아닌 클래스에 test를 썻는데 테스트 수행도 안되고 경고도 발생안함
#### 프로그램 요소를 매개변수로 전달할 마땅한 방법이 없다.

<br/>

## 애너테이션
- 애너테이션은 위 문제를 모두 해결한다.

<br/>

## marker 애너테이션
- marker 애너테이션이란 아무 매개변수 없이 단순히 대상에 마킹한다는 뜻
```java
/**
 * 테스트 메서드임을 선언하는 애너테이션이다.
 * 매개변수 없는 정적 메서드 전용이다.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface Test {
}
```
- `@Retention`, `@Target`은 메타 애너테이션
- `@Retention(RetentionPolicy.RUNTIME)` : `@Test`가 런타임에도 유지되야 한다는 표시
  - 생략하면 테스트 도구는 `@Test` 인식 못함
- `@Target(ElementType.METHOD)` : `@Test`가 반드시 메서드 선언에서만 사용돼야 한다고 알려줌
### 사용 프로그램
```java
public class Sample {
    @Test
    public static void m1() { }        // 성공해야 한다.
    @Test public static void m3() {    // 실패해야 한다.
        throw new RuntimeException("실패");
    }
    public static void m4() { }  // 테스트가 아니다.
    @Test public void m5() { }   // 잘못 사용한 예: 정적 메서드가 아니다.
    public static void m6() { }
    @Test public static void m7() {    // 실패해야 한다.
        throw new RuntimeException("실패");
    }
}
```
### 처리 프로그램
```java
public class RunTests {
    public static void main(String[] args) throws Exception {
        int tests = 0;
        int passed = 0;
        Class<?> testClass = Class.forName(args[0]);
        for (Method m : testClass.getDeclaredMethods()) {
            if (m.isAnnotationPresent(Test.class)) { // isAnnotationPresent는 실행할 메서드를 찾아줌
                tests++;
                try {
                    m.invoke(null);
                    passed++;
                } catch (InvocationTargetException wrappedExc) { // 테스트 메서드가 예외를 던질때 처리 (위 m3, m7 메서드)
                    Throwable exc = wrappedExc.getCause();
                    System.out.println(m + " 실패: " + exc);
                } catch (Exception exc) { // @Test를 잘못사용했다는뜻 (매개변수가 있는 메서드, 인스턴스 메서드.. 등)
                    System.out.println("잘못 사용한 @Test: " + m);
                }
            }
        }
        System.out.printf("성공: %d, 실패: %d%n",
                passed, tests - passed);
    }
}
```
- `@Test`는 대상 코드의 의미에 직접적인 영향을 주지 않고, 그 애너테이션에 관심 있는 도구에서 특별한 처리를 할 수 있게 추가 정보를 제공

<br/>

## 매개변수 하나를 받는 애너테이션 타입
```java
/**
 * 명시한 예외를 던져야만 성공하는 테스트 메서드용 애너테이션
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface ExceptionTest {
    Class<? extends Throwable> value();
}
```
- `Class<? extends Throwable> `는 Throwable을 확장한 클래스의 Class 객체라는 뜻이다. (모든 예외타입을 수용)
### 사용 프로그램
```java
// 매개변수 하나짜리 애너테이션을 사용한 프로그램
public class Sample2 {
    @ExceptionTest(ArithmeticException.class)
    public static void m1() {  // 성공해야 한다.
        int i = 0;
        i = i / i;
    }
    @ExceptionTest(ArithmeticException.class)
    public static void m2() {  // 실패해야 한다. (다른 예외 발생)
        int[] a = new int[0];
        int i = a[1];
    }
    @ExceptionTest(ArithmeticException.class)
    public static void m3() { }  // 실패해야 한다. (예외가 발생하지 않음)
}
```
### 처리 프로그램
```java
//위 marker 애너테이션 main메서드 다음처럼 수정
if (m.isAnnotationPresent(ExceptionTest.class)) {
     tests++;
     try {
         m.invoke(null);
         System.out.printf("테스트 %s 실패: 예외를 던지지 않음%n", m);
     } catch (InvocationTargetException wrappedEx) {
         Throwable exc = wrappedEx.getCause();
         Class<? extends Throwable> excType =
                 m.getAnnotation(ExceptionTest.class).value();
         if (excType.isInstance(exc)) { // 올바른 예외를 던지는지 확인
             passed++;
         } else {
             System.out.printf(
                     "테스트 %s 실패: 기대한 예외 %s, 발생한 예외 %s%n",
                     m, excType.getName(), exc);
         }
     } catch (Exception exc) {
     System.out.println("잘못 사용한 @ExceptionTest: " + m);
 }
}    
```

<br/>

## 배열 매개변수를 받는 애너테이션 타입

- 예외를 여러개 명시하고 그 중 하나가 발생하면 성공하게 처리

```java
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface ExceptionTest {
    Class<? extends Exception>[] value();
}
```

### 사용 프로그램
```java
// 이 변형은 원소 하나짜리 매개변수를 받는 애너테이션도 처리할 수 있다. (241쪽 Sample2와 같음)
    @ExceptionTest(ArithmeticException.class)
    public static void m1() {  // 성공해야 한다.
        int i = 0;
        i = i / i;
    }
    @ExceptionTest(ArithmeticException.class)
    public static void m2() {  // 실패해야 한다. (다른 예외 발생)
        int[] a = new int[0];
        int i = a[1];
    }
    @ExceptionTest(ArithmeticException.class)
    public static void m3() { }  // 실패해야 한다. (예외가 발생하지 않음)

    // 코드 39-7 배열 매개변수를 받는 애너테이션을 사용하는 코드 (242-243쪽)
    @ExceptionTest({ IndexOutOfBoundsException.class,
                     NullPointerException.class })
    public static void doublyBad() {   // 성공해야 한다.
        List<String> list = new ArrayList<>();

        // 자바 API 명세에 따르면 다음 메서드는 IndexOutOfBoundsException이나
        // NullPointerException을 던질 수 있다.
        list.addAll(5, null);
    }
}
```
### 처리 프로그램
```java
if (m.isAnnotationPresent(ExceptionTest.class)) {
    tests++;
    try {
        m.invoke(null);
        System.out.printf("테스트 %s 실패: 예외를 던지지 않음%n", m);
    } catch (Throwable wrappedExc) {
        Throwable exc = wrappedExc.getCause();
        int oldPassed = passed;
        Class<? extends Throwable>[] excTypes =
                m.getAnnotation(ExceptionTest.class).value(); // 주목할부분
        for (Class<? extends Throwable> excType : excTypes) { // 주목할부분
            if (excType.isInstance(exc)) {
                passed++;
                break;
            }
        }
        if (passed == oldPassed)
            System.out.printf("테스트 %s 실패: %s %n", m, exc);
    }
}
```

<br/>

## 반복 가능한 애너테이션 타입
- __여러 개 값을 받는 애너테이션을 구현하는 또 다른 방법이다.__
- 배열 매개변수 사용하는 대신 여러 번 달 수 있는 `@Repeatable` 메타애너테이션을 사용
### 주의할점
- `@Repeatable`을 단 애너테이션을 반환하는 컨테이너 애너테이션을 하나 더 정의하고 `@Repeatable`에 이 컨테이너 애너테이션의 class객체를 매개변수로 전달해야한다.
- 컨테이너 애너테이션은 내부 애너테이션 타입의 배열을 반환하는 value 메서드를 정의해야한다.
- 컨테이너 애너테이션 타입에는 적절한 보존정책(`@Retention`)과 적용대상(`@Target`)을 명시해야함.
```java
// 반복 가능한 애너테이션 타입
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Repeatable(ExceptionTestContainer.class)
public @interface ExceptionTest {
    Class<? extends Throwable> value();
}
// 반복 가능한 애너테이션의 컨테이너 애너테이션
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface ExceptionTestContainer {
    ExceptionTest[] value();
}
```
### 사용 프로그램
```java
@ExceptionTest(IndexOutOfBoundsException.class)
@ExceptionTest(NullPointerException.class)
public static void doublyBad() {
    List<String> list = new ArrayList<>();

    // 자바 API 명세에 따르면 다음 메서드는 IndexOutOfBoundsException이나
    // NullPointerException을 던질 수 있다.
    list.addAll(5, null);
}
```
### 처리 프로그램
```java
if (m.isAnnotationPresent(ExceptionTest.class)
                    || m.isAnnotationPresent(ExceptionTestContainer.class)) { // 주목할부분
    tests++;
    try {
        m.invoke(null);
        System.out.printf("테스트 %s 실패: 예외를 던지지 않음%n", m);
    } catch (Throwable wrappedExc) {
        Throwable exc = wrappedExc.getCause();
        int oldPassed = passed;
        ExceptionTest[] excTests =
                m.getAnnotationsByType(ExceptionTest.class); // 주목할부분
        for (ExceptionTest excTest : excTests) {  // 주목할부분
            if (excTest.value().isInstance(exc)) {
                passed++;
                break;
            }
        }
        if (passed == oldPassed)
            System.out.printf("테스트 %s 실패: %s %n", m, exc);
    }
  }
```
- 반복가능 애너테이션을 여러개 달면 하나만 달렸을 때와 구분 위해 해당 컨테이너 애너테이션 타입적용
- `isAnnotationPresent`는 반복가능애너테이션(`ExceptionTest`)과 해당 컨테이너 애너테이션(`ExceptionTestContainer`)을 구분한다. -> 따로따로 확인해야함
- 반면 `getAnnotationsByType`는 구분하지 않아 모두 가져온다.
